### DISCUSSION TITLE
Cannot get a license expiry value for trial licenses

### DISCUSSION THREAD

**Discussion by fmg-pete** [keygen-sh/community#93] [Q&A] [answered]
We are trying to implement a 14 day trial license, I've fudged the timestamp in the policy, and that works as expected. If I create a license using that policy in the Keygen dashboard, the expiry date is null until the first validation, then it uses the timestamp of the validation and adds 14 days to it to determine the expiry. Again, that's as I'd expect. 

The problem I have, is that from an app perspective, there is no way to get the calculated expiry date at validation time. I store the company name and license key in the Windows registry, but the expiry date is always null at this point. I tried adding a metadata key/value pair to the policy with the 14 day offset in seconds, with the hope of adding that to the validation timestamp and saving THAT value in the registry, but no luck. None of those values appear in the validation response. I considered adding the same key to the actual License as metadata, but that seems really klunky to me.

Does anyone have a solution for this?

**Comment by ezekg** [comment]
This behavior is due to your policy's expiration basis, which is documented here: https://keygen.sh/docs/api/policies/#policies-object-attrs-expirationBasis

It sounds like you're using a `FROM_FIRST_VALIDATION` expiration basis, so the expiry is set _after_ the first validation event occurs, not during. So the expiry will still be null for the first validation (because its set asynchronously, after the validation), but _should_ be non-null for subsequent validations.

Also worth noting that policy metadata is separate from license metadata — there's no inheritance i.r.t metadata. You may be able to use entitlements to store inherited metadata, if that works for you.

Hope all that helps? Let me know.

**Reply by fmg-pete** [reply]
Thanks for thee lightning fast reply! 

I did look at the policy expiration, but the only one that would work is "From Creation". This could be problematic for a 14 day trial, as it might be 2 days after I create the license that the business sends it to the user, and then say 1 or 2 days before the user gets around to installing it and using the app for the first time. Their 14 day trial becomes a nine-and-a-half day trial - pretty much be the same as me just punching in a date in 2 weeks into the license at creation time and hoping for the best. I need the end user to get 14 days from the time they validate, or first ping, or something... and then stash that date in the app where they can see it ( like an "About" view, "License expires in 3 days". 

Is there another way I can do this?

**Reply by ezekg** [reply, marked as answer]
Since the expiry is set asynchronously after validation, making a request to [retrieve the license](https://keygen.sh/docs/api/licenses/#licenses-retrieve) after validation may be a suitable workaround here. Another solution would be to simply show an interim value in the about page until next validation, at which time the expiry will likely have been set.

**Reply by fmg-pete** [reply]
Thanks again, that last sentence was the silver bullet! :)
